client/starwhale/api/_impl/dataset/model.py

```diff
@@ -0,0 +1,665 @@
+
+    def remove(self, force: bool = False) -> None:
+        ok, reason = self.__core_dataset.remove(force)
+        if not ok:
+            raise RuntimeError(f"failed to remove dataset: {reason}")
+
+    def recover(self, force: bool = False) -> None:
+        ok, reason = self.__core_dataset.recover(force)
+        if not ok:
+            raise RuntimeError(f"failed to recover dataset: {reason}")
+
+    def summary(self) -> t.Optional[DatasetSummary]:
+        return self._summary
+
+    def history(self) -> t.List[t.Dict]:
+        return self.__core_dataset.history()
+
+    def close(self) -> None:
+        self.__keys_cache = set()
+        for _loader in self.__data_loaders.values():
+            if not _loader:
+                continue  # pragma: no cover
+
+            _loader.tabular_dataset.close()
+
+        if self._row_writer:
+            self._row_writer.close()
+
+        # TODO: flush raw data into disk
+
+    def diff(self, cmp: Dataset) -> t.Dict:
+        return self.__core_dataset.diff(cmp.uri)
+
+    def info(self) -> t.Dict[str, t.Any]:
+        return self.__core_dataset.info()
+
+    def head(self, n: int = 3, show_raw_data: bool = False) -> t.List[t.Dict]:
+        # TODO: render artifact in JupyterNotebook
+        return self.__core_dataset.head(n, show_raw_data)
+
+    def to_pytorch(self) -> t.Any:
+        raise NotImplementedError
+
+    def to_tensorflow(self) -> t.Any:
+        raise NotImplementedError
+
+    @_check_readonly
+    @_forbid_handler_build
+    def __setitem__(
+        self, key: t.Union[str, int], value: t.Union[DataRow, t.Tuple]
+    ) -> None:
+        # TODO: tune the performance of getitem by cache
+        self._trigger_icode_build = True
+        if not isinstance(self.__core_dataset, StandaloneDataset):
+            raise NoSupportError(
+                f"setitem only supports for standalone dataset: {self.__core_dataset}"
+            )
+
+        _row_writer = self._get_row_writer()
+
+        if not isinstance(key, (int, str)):
+            raise TypeError(f"key must be str or int type: {key}")
+
+        if isinstance(value, DataRow):
+            value.index = key
+            row = value
+        elif isinstance(value, (tuple, list)):
+            if len(value) == 2:
+                data, annotations = value
+            elif len(value) == 3:
+                _, data, annotations = value
+            else:
+                raise ValueError(f"{value} cannot unpack")
+
+            row = DataRow(index=key, data=data, annotations=annotations)
+        else:
+            raise TypeError(f"value only supports tuple or DataRow type: {value}")
+
+        if key not in self.__keys_cache:
+            self.__keys_cache.add(key)
+            _item = self._getitem(key, skip_fetch_data=True)
+            if _item is None or len(_item) == 0:
+                self._rows_cnt += 1
+
+        _row_writer.update(row)
+
+    def _get_row_writer(self) -> RowWriter:
+        with self._writer_lock:
```

Comment:
the code below will make locking only happens when self._row_writer is none
```python
def _get_row_writer(self) -> RowWriter:
        if self._row_writer:
            return self._row_writer
        with self._writer_lock:

```

Reply:
updated